### PR TITLE
Adds dbt artifact URLs to DbtCloudRunJob

### PR DESCRIPTION
**Pull request by desertaxle**
## Summary

Adds the creation of link artifacts to capture the dbt Cloud job run artifacts in the Prefect Cloud console and also includes the links to dbt artifacts in the return value of the run method in the DbtCloudRunJob task.

## Changes

Adds artifacts and adds a key to the existing return type. Also updates exceptions raised in dbt cloud utils.

## Importance

Increases the discoverability and usability of dbt Cloud artifacts from job runs.

## Checklist

This PR:

- [x] adds new tests (if appropriate)
- [x] adds a change file in the `changes/` directory (if appropriate)
- [x] updates docstrings for any new functions or function arguments, including `docs/outline.toml` for API reference docs (if appropriate)

### CODE REVIEW

**Comment by sungchun12** [comment]
Happy to test this in the wild before it's merged!

**Comment by desertaxle** [comment]
Thanks @sungchun12! Do you have a sample dbt Cloud job that we could call from a Prefect flow?

**Comment by anna-geller** [comment]
@desertaxle just tested it - it works fine, but the artifacts didn't show up:
image

However, after changing `(create_link_artifact(link=link) for link in artifact_links)` to `create_link_artifact(link=artifact_links[1])` in the next flow run, it works fine:
image

Basically, prefect.backend.artifacts.create_link_artifact allows to only show one URL, but if we want to show multiple links in the artifact, then markdown artifact will probably work better:
```python
        if wait_for_job_run_completion:

            artifact_links = create_run_artifact_links(
                account_id=account_id, run_id=run["id"]
            )
           markdown = ""

            for url in artifact_links:
                markdown += f"- {url} \n"
            
            create_markdown_artifact(markdown)
```
Here is how it looked like in the flow run:
image

**Comment by desertaxle** [comment]
@anna-geller Thanks for testing! That's a good call on generating a markdown artifact to hold multiple links. One question before I make that change, are you able to see the artifacts in your browser when you click on those links?

**Comment by sungchun12** [comment]
@desertaxle I do have an example dbt Cloud job. Are you suggesting I test it like Anna above or did you want permission access to invoke the job from prefect during your testing?

**Comment by anna-geller** [comment]
> @anna-geller Thanks for testing! That's a good call on generating a markdown artifact to hold multiple links. One question before I make that change, are you able to see the artifacts in your browser when you click on those links?

@desertaxle I can open all of those URLs just fine in my browser, or save the files to my laptop: 
image
image

**Comment by anna-geller** [comment]
@desertaxle actually, the last link is always empty - is it because I haven't used any catalog in my dbt models? not sure what this catalog file is about but it's always empty for me, perhaps it can be skipped for the sake of artifact?
image

**Comment by desertaxle** [comment]
> @desertaxle actually, the last link is always empty - is it because I haven't used any catalog in my dbt models? not sure what this catalog file is about but it's always empty for me, perhaps it can be skipped for the sake of artifact? image

@sungchun12 Do you know why we wouldn't have a `catalog.json` artifact from a dbt Cloud job run? Is there a way to determine if a `catalog.json` was generated before we create the Prefect artifact?

**Comment by anna-geller** [comment]
> > @desertaxle actually, the last link is always empty - is it because I haven't used any catalog in my dbt models? not sure what this catalog file is about but it's always empty for me, perhaps it can be skipped for the sake of artifact? image
> 
> @sungchun12 Do you know why we wouldn't have a `catalog.json` artifact from a dbt Cloud job run? Is there a way to determine if a `catalog.json` was generated before we create the Prefect artifact?

@desertaxle good question, it looks like it's populated when someone uses the "dbt docs generate" command - let me try to add that command and see if file gets populated https://docs.getdbt.com/reference/artifacts/catalog-json

**Comment by anna-geller** [comment]
@desertaxle ok, confirmed. When adding "dbt docs generate", this file is no longer empty, so we need to render all three URLs and we can perhaps explain in the docstring that this will generate 3 links and clarify which dbt command each link represents:
image

**Comment by sungchun12** [comment]
@desertaxle @anna-geller I confirm that you need to run `dbt docs generate` to populate the `catalog.json`
image


The same applies for `sources.json` requiring `dbt source freshness` in my earlier recommendation to add that as an artifact.

I recommend that the prefect task dynamically surfaces the artifacts if they exist in the dbt cloud job run.

**Comment by desertaxle** [comment]
@sungchun12 I made a change to pull all the dbt artifacts via the [list artifacts endpoint](https://docs.getdbt.com/dbt-cloud/api-v2#operation/listArtifactsByRunId) so the artifacts will be dynamic based on the run. @anna-geller if that looks good in the Prefect UI then I think this should be good for full review.

**Comment by anna-geller** [comment]
Something went wrong in the last commit:
image
dbt_cloud_utils.py, line 188, in list_run_artifact_links - seems to return None?

**Comment by anna-geller** [comment]
@desertaxle 

I made a tiny change and this works:
```python
        run = trigger_job_run(
            account_id=account_id,
            job_id=job_id,
            cause=cause,
            additional_args=additional_args,
            token=token,
        )
        if wait_for_job_run_completion:

            artifact_links = []
            try:
                job_run_result = wait_for_job_run(
                    account_id=account_id,
                    run_id=run["id"],
                    token=token,
                    max_wait_time=max_wait_time,
                )
                artifact_links = list_run_artifact_links(
                    account_id=account_id, run_id=run["id"], token=token
                )

                markdown = f"Artifacts for dbt Cloud run {run['id']} of job {job_id}\n"
                for link, name in artifact_links:
                    markdown += f"- [{name}]({link})\n"
                from prefect.backend.artifacts import create_markdown_artifact
                create_markdown_artifact(markdown)
            except DbtCloudListArtifactsFailed as err:
                self.logger.warn(
                    f"Unable to retrieve artifacts generated by dbt Cloud job run: {err}"
                )

            job_run_result["artifact_urls"] = [link for link, _ in artifact_links]

            return job_run_result

        else:
            return run
```

image

File: src/prefect/tasks/dbt/dbt.py [hunk 1]
```diff
@@ -407,12 +413,6 @@ def run(
                 for link, name in artifact_links:
                     markdown += f"- [{name}]({link})\n"
```
Comment (anna-geller) [inline src/prefect/tasks/dbt/dbt.py:415]: @desertaxle you still need to actually call the create_markdown_artifact method:
```python
                markdown = f"Artifacts for dbt Cloud run {run['id']} of job {job_id}\n"
                for link, name in artifact_links:
                    markdown += f"- [{name}]({link})\n"
                from prefect.backend.artifacts import create_markdown_artifact
                create_markdown_artifact(markdown)
```

File: changes/pr5135.yaml [hunk 2]
```diff
@@ -0,0 +1,2 @@
+enhancement:
+  - "Adds artifacts to DbtCloudRunJob task - [#5135](https://github.com/PrefectHQ/prefect/pull/5135)"
```
Comment (madkinsz) [inline changes/pr5135.yaml:2]:
```suggestion
  - "Add artifact publishing to `DbtCloudRunJob` task - [#5135](https://github.com/PrefectHQ/prefect/pull/5135)"
```

File: src/prefect/tasks/dbt/dbt_cloud_utils.py [hunk 3]
```diff
@@ -12,6 +14,62 @@
...
+#   https://docs.getdbt.com/dbt-cloud/api-v2#operation/getArtifactsByRunId
+__DBT_CLOUD_GET_RUN_ARTIFACT_ENDPOINT_V2 = (
+    "https://cloud.getdbt.com/api/v2/accounts/{accountId}/runs/{runId}/artifacts/{path}"
+)
+
+dbt_cloud_artifact_paths = ("manifest.json", "run_results.json", "catalog.json")
+
+
+class DbtCloudBaseException(Exception):
+    """Base excpetion for all dbt Cloud Excpetions"""
```
Comment (madkinsz) [inline src/prefect/tasks/dbt/dbt_cloud_utils.py:33]:
```suggestion
    """Base exception for all dbt Cloud errors"""
```

**Review by madkinsz — CHANGES_REQUESTED** [review]
Some minor changes

Otherwise, LGTM!

**Review by madkinsz — APPROVED** [review]

**Pull request merged**